### DISCUSSION TITLE
Is @InputArgument necessary for each type parameter of a query?

### DISCUSSION THREAD

**Discussion by albertogdd** [Netflix/dgs-framework#529] [Q&A] [answered]
From the [docs](https://netflix.github.io/dgs/datafetching/#using-inputargument) I understand that `@InputArgument` is necessary with each parameter in a query/mutation, but I tried to have the data fetchers without it and it also works:
```java
@DgsQuery
public Long countMovies(
     Long countryId,
     String movieTitle,
     Long director,
     Genre genre
) {
return null;
}
```
Is it only necessary for Input types? Or I'm missing something here. In the Java project example you guys have all parameters are tagged with `@InputArgument`.

The funny thing is that if I put `@InputArgument` on each of them, this exception pops up with the optional parameter `Long director`:
```java
@DgsQuery
public Long countMovies(
     @InputArgument Long countryId,
     @InputArgument String movieTitle,
     @InputArgument Long director,
     @InputArgument Genre genre
) {
return null;
}
```
```
com.netflix.graphql.dgs.exceptions.DgsInvalidInputArgumentException: Specified type 'class java.lang.Long' is invalid. Found java.lang.String instead.
```

**Comment by paulbakker** [comment, marked as answer]
When no annotation is found, there is some fallback logic that tries to match the arguments by name. This only works if the compiler settings preserve argument names, so it's not recommended to rely on it.

I'm not sure why in your example it didn't work with `@InputArgument`. Assuming the schema and query were defined correctly that should definitely work.
We did make some significant changes in the latest release candidate (`4.6.0-rc.5`), which improves how input arguments are processed. Might be worth trying with that release.
